### PR TITLE
Fix tensor initialiser and function return types

### PR DESCRIPTION
**Pull request by jppelteret**
Two small patches:
1. In the tensor constructors, use a cast to convert the underlying data in the tensor when the input initialiser has a different `NumberType`. This is effectively what we do [here as well](https://github.com/dealii/dealii/blob/master/include/deal.II/base/tensor.h#L829).
2. Correctly initialise some return values using our nifty internal template struct.

### CODE REVIEW

**Comment by bangerth** [comment]
/run-tests
OK to merge once the tester is happy.

**Comment by masterleinad** [comment]
The failing test needs to be fixed:
> 171/3275 Test #177: physics/elasticity-kinematics_02.debug ...............................***Failed    1.56 sec
Test physics/elasticity-kinematics_02.debug: BUILD
===============================   OUTPUT BEGIN  ===============================
physics/elasticity-kinematics_02.debug: BUILD failed. Output:
[ 20%] Generating elasticity-kinematics_02.debug/interrupt_guard.cc
Scanning dependencies of target elasticity-kinematics_02.debug
[ 40%] Building CXX object CMakeFiles/elasticity-kinematics_02.debug.dir/elasticity-kinematics_02.cc.o
In file included from /home/bob/source/include/deal.II/physics/elasticity/kinematics.h:20:0,
                 from /home/bob/source/tests/physics/elasticity-kinematics_02.cc:20:
/home/bob/source/include/deal.II/base/symmetric_tensor.h: In instantiation of â€˜dealii::SymmetricTensor<rank, dim, Number>::SymmetricTensor(const dealii::SymmetricTensor<rank, dim, OtherNumber>&) [with OtherNumber = double; int rank = 2; int dim = 3; Number = dealii::VectorizedArray<double>]â€™:
/home/bob/source/tests/physics/elasticity-kinematics_02.cc:75:153:   required from here
/home/bob/source/include/deal.II/base/symmetric_tensor.h:1002:13: error: no matching function for call to â€˜dealii::VectorizedArray<double>::VectorizedArray(const value_type&)â€™
     data[i] = static_cast<typename base_tensor_type::value_type>(initializer.data[i]);

**Comment by jppelteret** [comment]
Thanks, I'm on it already!

File: include/deal.II/base/numbers.h [hunk 1]
```diff
@@ -397,7 +397,7 @@ namespace internal
   template <typename T>
   struct NumberType
   {
-    static DEAL_II_CUDA_HOST_DEV T value (const T &t)
+    static DEAL_II_CUDA_HOST_DEV const T &value (const T &t)
```
Comment (jppelteret) [inline include/deal.II/base/numbers.h]: @Rombur If this PR gets merged as is, and `CUDA` breaks then please keep this change in mind!

**Comment by jppelteret** [comment]
Ok, I think that I've fixed the broken test but this needs a re-review (and verification by the tester that I've not duffed something else up). It turns out that `VectorizedArray` doesn't mix with `static_cast`, which is a pity. I think that I've worked around this and I think that I have a plan to work around it my yet-to-be-submitted work.

I've also taken the liberty to extend the various specialisations of the `NumberType` struct, in so that we now ensure that if the input and output types are the same then we return the input object by reference as opposed to making a copy of it.

File: include/deal.II/base/numbers.h (hunk 1 above)
Comment (masterleinad) [inline include/deal.II/base/numbers.h]: I can compile with CUDA and run all `cuda` tests successfully.

Comment (jppelteret) [inline include/deal.II/base/numbers.h]: Thanks for confirming this!

**Comment by jppelteret** [comment]
Relates #4958

**Pull request merged**